### PR TITLE
Fix issue with components counter for cached resources

### LINKED ISSUE #7227: Do not use cached version of the collection if component counts has changed
**Describe the bug (REQUIRED)**

- Bob builds a collection and counts components it uses
- cache it
- Developer make changes in a go file on disk referenced by the collection but not in the collection itself
- Bob doesn’t take these changes into account because for the collection itself nothing has changed
- Build system takes this collection from cache

**Expected behavior (REQUIRED)**
Collection should be rebuilt if something has changed in its dependencies

REPO:
[issue7227-collections-cache.zip](https://github.com/defold/defold/files/10277511/issue7227-collections-cache.zip)
1. build project using 
```
java -jar ~/Downloads/bob.jar --platform=x86_64-macos resolve clean build --archive bundle --variant debug --resource-cache-local ~/Downloads/cache -v -bo ~/Downloads/build
```
2. Run project - it works.
3. Add one more `factory` component in `one.go`
4. Build again.
5. Run project - it doesn't work
>ERROR:GAMESYS: Can not create more factory components since the buffer is full (2). See 'factory.max_count' in game.project

### PR DESCRIPTION
**Pull request by AGulev**
Now Bob considers changes in a linked GameObject when deciding to take a collection from the cache or rebuild it.

Fix https://github.com/defold/defold/issues/7227

## PR checklist

* [x] Code
	* [x] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/CollectionBuilderTest.java [hunk 1]
```diff
@@ -671,6 +685,59 @@ public void testInstanceComponentCounterInFactory() throws Exception {
...
+     * Test that the component counter ignores collections with dynamic factories
+     * Structure:
+     * - go
+     *   - factory
+     *      - collection
+     *          - sprite
+     * @throws Exception
+     */
+    @Test
+    public void testInstanceComponentCounterForDynamicFactory() throws Exception {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/CollectionBuilderTest.java:698]: new test for components factory when collection has a dynamic factory

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/ParseUtil.java [hunk 2]
```diff
@@ -204,6 +204,18 @@ public Message parse(byte[] content) throws InvalidProtocolBufferException {
...
+        parseMap.put("compcount_col", new IParser() {
+            @Override
+            public Message parse(byte[] content) throws InvalidProtocolBufferException {
+                return null;
+            }
+        });
+        parseMap.put("compcount_go", new IParser() {
+            @Override
+            public Message parse(byte[] content) throws InvalidProtocolBufferException {
+                return null;
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/ParseUtil.java:216]: these are not a Proto Messsage

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/CollectionBuilder.java [hunk 3]
```diff
@@ -113,169 +91,7 @@ private void createGeneratedResources(Project project, CollectionDesc.Builder bu
             CollectionDesc.Builder subCollectionBuilder = CollectionDesc.newBuilder();
             ProtoUtil.merge(collectionResource, subCollectionBuilder);
 
-            createGeneratedResources(project, subCollectionBuilder);
-        }
-    }
-
-    // TODO: do we really need this?
-    // This method is never called. Is it mistake?
-    private int buildEmbedded(IResource input, CollectionDesc.Builder builder, Task<Void> task, int embedIndex) throws IOException, CompileExceptionError {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/CollectionBuilder.java:122]: unused function

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/CollectionBuilder.java [hunk 4]
```diff
@@ -294,37 +110,49 @@ public Task<Void> create(IResource input) throws IOException, CompileExceptionEr
...
+        Map<Long, IResource> uniqueResources = new HashMap<>();
+        createGeneratedResources(this.project, builder, uniqueResources);
+        
+        List<Task<?>> embedTasks = new ArrayList<>();
         for (long hash : uniqueResources.keySet()) {
             IResource genResource = uniqueResources.get(hash);
-
             taskBuilder.addOutput(genResource);
-
-            if (productsOfThisTask.contains(hash)) {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/CollectionBuilder.java:319]: `productsOfThisTask` is always equal to `uniqueResources` (see line 133-134)
I don't see the reson to have `productsOfThisTask`

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/TimeProfiler.java [hunk 5]
```diff
@@ -51,7 +51,6 @@ public class TimeProfiler {
      * Helper class that contains profiling data and represents a linked list of scopes hierarchy.
      */
     private static class ProfilingScope {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/TimeProfiler.java:53]: I made changes that broke TimeProfiler, so I fixed it here.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/TimeProfiler.java [hunk 6]
```diff
@@ -51,7 +51,6 @@ public class TimeProfiler {
...
-        public String name;
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/TimeProfiler.java:54]: we don't really use it except 2 cases, I changed to more generic way

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/ComponentsCounter.java [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+            }
+        }
+
+        public byte[] toByteArray() throws IOException {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(128 * 16);
+            ObjectOutputStream os = new ObjectOutputStream(bos);
+            os.writeObject(this);
+            os.close();
+            return bos.toByteArray();
+        }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/ComponentsCounter.java:117]: For debuggability, would it be possible to store them as text? (I assume there's some Java class that helps with this)

**Review by JCash — DISMISSED** [review]
Looks good (hard to see any actual counting issues)

Apart from minor question about the addInput, I also wonder about the debuggability of the count formats.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Task.java [hunk 8]
```diff
@@ -76,6 +76,13 @@ public TaskBuilder<T> addInput(IResource input) {
             return this;
         }
 
+        public TaskBuilder<T> addInputIfUnique(IResource input) {
+            if (!task.inputs.contains(input)){
+                task.inputs.add(input);
+            }
+            return this;
+        }
```
[the remark this answers is not in the document]
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Task.java:84]: Makes sense, I'll fix it
thx

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/ComponentsCounter.java (hunk 7 above)
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/ComponentsCounter.java:117]: I thin it's easier to cover with extra logging (esp taking into account that we have collections which counts embeded objects themself and can't take date from embeded objects' `compcount_go`)

**Pull request merged**